### ISSUE TITLE
[Android][css-layout] Equal-height columns 

### ISSUE THREAD

**Issue by corbt** [facebook/react-native#2724]
**UPDATE**: For those just arriving, see my [next comment](#issuecomment-140491988) on this issue for a simplified and reproducible test case.

--------------------------------------
I've been playing around with React Native on Android. I'm implementing a toy Hacker News reader to learn about the API.

I'm running into an issue that appears to be related to either the flexbox model or React Native's implementation of it. I'm the first to admit that I'm not a flexbox expert so there may be an obvious solution to my problem that I'm not aware of.

Essentially, I have two columns, each wrapped inside a `TouchableHighlight`. I want the heights to be equal (and I don't know ahead of time which will be taller). I've found that if I manually set the height of the parent view, as seen below, everything is calculated correctly. However, if I allow the parent View to be calculated based on the heights of its two children, the `alignItems: 'stretch'` stops working and I'm not able to center the views vertically and make them the same height.

I wasn't totally sure whether to open this as an issue here or post it on Stack Overflow, but since I'm not sure whether it would be seen there (and because it's possibly related to a bug in the Flexbox implementation) I thought I'd ask here.

```jsx
<View style={{borderBottomWidth: 1, borderBottomColor: '#FFB65D', flexDirection: 'row', alignItems: 'stretch', height: 100}}>
  <TouchableHighlight style={{flex: 1}}>
    <View style={{flex: 1, padding: 10, backgroundColor: 'white', justifyContent: 'center'}}>
      <Text style={{fontSize: 15, color: 'black'}}>{this.props.story.title}</Text>
      <View style={{flexDirection: 'row', color: 'gray'}}>
        <Text>{this.props.story.points} points</Text>
        {separator}
        <Text>{this.props.story.user}</Text>
        {separator}
        <Text>{this.props.story.time_ago}</Text>
      </View>
    </View>
  </TouchableHighlight>
  <TouchableHighlight>
    <View style={{flex: 1, backgroundColor: '#FFCC8D', alignItems: 'center', justifyContent: 'center', flexDirection: 'column', padding: 5}}>
      <Text style={{fontSize: 30}}>{this.props.story.comments_count}</Text>
      <Text style={{fontSize: 10}}>COMMENTS</Text>
    </View>
  </TouchableHighlight>
</View>
```

**Comment by mkonicek** [comment]
Our Flexbox implementation is consistent with web: https://github.com/facebook/css-layout

Does this help? https://css-tricks.com/snippets/css/a-guide-to-flexbox/

Also, please use Stack Overflow for asking questions and Github issues for reporting bugs:
http://stackoverflow.com/questions/tagged/react-native

**Comment by corbt** [comment]
I think I may have found a couple of inconsistencies that are related to my problem. I've created what I believe are equivalent layouts in React Native and HTML/CSS that appear to demonstrate different flexbox behavior. Am I just crazy or is this a bug?

## HTML
JSFiddle [here](https://jsfiddle.net/awtLxLpm/)

## React Native
```jsx
<View style={{flexDirection: 'row', alignItems: 'stretch', width: 150}}>
  <View style={{backgroundColor: 'red'}}>
    <View style={{backgroundColor: 'green'}}>
      <Text style={{backgroundColor: 'purple', alignSelf: 'center'}}>
        Column 1
      </Text>
    </View>
  </View>
  <Text style={{backgroundColor: '#B2DBFF', height: 150}}>Column 2</Text>
</View>
```

Output on my Android device:
![device](https://cloud.githubusercontent.com/assets/176426/9885921/a27e94e6-5bdf-11e5-9b9d-3b1d3405c19e.png)

## Differences
In React Native the intermediary green `div` doesn't expand to fill the column vertically, and the "Column 1" text is not centered vertically either. On my machine (OS X/Chrome 45) the HTML version does.

**Comment by brentvatne** [comment]
@corbt - a closer example to this is:

```javascript
      <View style={{flexDirection: 'row', alignItems: 'stretch', width: 150}}>
        <View style={{backgroundColor: 'red'}}>
          <View style={{backgroundColor: 'green'}}>
            <View style={{backgroundColor: 'purple', alignSelf: 'center'}}>
              <Text>
                Column 1
              </Text>
            </View>
          </View>
        </View>
                
        <View style={{backgroundColor: '#B2DBFF', height: 150}}>
          <Text>Column 2</Text>
        </View>
      </View>
```

You should apply the Flex attributes to `View` and not `Text`. That said, this still doesn't work :P

**Comment by brentvatne** [comment]
The bug appears to be with `flexDirection: 'row'` as it works properly without it (albeit on the incorrect axis):

```
<View style={{alignItems: 'stretch', width: 150}}>
  <View style={{backgroundColor: 'red'}}>
    <View style={{backgroundColor: 'green'}}>
      <View style={{backgroundColor: 'purple', alignSelf: 'center'}}>
        <Text>
          Column 1
        </Text>
      </View>
    </View>
  </View>

  <View style={{backgroundColor: '#B2DBFF', height: 150}}>
    <Text>Column 2</Text>
  </View>
</View>
```

**Comment by mkonicek** [comment]
@brentvatne Interesting, `flexDirection: 'row'` should be the default so removing it should have no effect.

This indeed seems inconsistent with web: e.g. `alignSelf: 'center'` should center the element vertically (cross axis) according to https://css-tricks.com/snippets/css/a-guide-to-flexbox/

I'm quite surprised that the green div is [stretched in Chrome](https://jsfiddle.net/awtLxLpm/) - looks like `alignItems: 'stretch'` propagates two levels down.

@lucasr, @vjeux is [css-layout](https://github.com/facebook/css-layout) consistent with the web?

**Comment by corbt** [comment]
Yes, I tested with the nested Text as in your correction as well, but since
the behavior was identical either way I posted the simpler one. Glad to
hear I'm not crazy though.

**Comment by lucasr** [comment]
@mkonicek COLUMN is current the default in css-layout for Android.

**Comment by lucasr** [comment]
@corbt Thanks for the report! I suggest you to file a separate issue in css-layout with a test case for the web. css-layout has a ton of unit tests that ensure we match Chrome's flexbox implementation. This seems to be an incosistency not covered by our existing tests.

**Comment by mkonicek** [comment]
Looks like the default `flex-direction` on the web is `row` while it's `column` on Android: https://jsfiddle.net/dvb1q684/1/

**Comment by mkonicek** [comment]
And the default for `align-items` is `stretch` which explains the second inconsistency - the green div is stretched: https://developer.mozilla.org/en-US/docs/Web/CSS/align-items

**Comment by corbt** [comment]
Yeah I can confirm that with the default styles as set by the css-layout project the web version turns out looking the same as the Android version. So this may not be a bug. Still looking into how to accomplish my original goal with Flexbox though.

**Comment by mkonicek** [comment]
Here are the defaults in css-layout:
- `flexDirection` - `column`: https://github.com/facebook/css-layout/blob/21202854676e7f3df0b73634bbc5beeb352fa281/src/Layout.js#L272
- `alignItems` - `stretch`, not sure yet why the green div is not stretched on Android: https://github.com/facebook/css-layout/blob/master/src/Layout.js#L238

**Comment by mkonicek** [comment]
It's possible to achieve the same behavior on web and using css-layout although different defaults might be confusing. We need to decide whether `flexDirection: column` is a good default for React Native.

Regarding "Equal-height columns", I don't know how to do this off the top of my head and would suggest asking on Stack Overflow, it's a question specific to Flexbox.

You basically want the height to be the max of the heights of the two columns. I think if you tell children to stretch to the size of the parent and tell the parent to match the size of the children, you've essentially created a loop and there's no clear way for the layout algorithm to resolve it.

If you had to calculate the layout by hand, would you be able to do it without measuring each of the columns first and then calculating `max(width(column1, column2))`?

**Comment by mkonicek** [comment]
OK, I asked @vjeux and the divergence is a deliberate decision: https://github.com/facebook/css-layout#default-values

**Comment by brentvatne** [comment]
@mkonicek - we should link to that in the Flexbox docs

**Comment by mkonicek** [comment]
I think the green view should still be stretched though:

    <View style={{backgroundColor: 'red'}}>
        <View style={{backgroundColor: 'green'}}>

Can you file an issue on css-layout, link to your Stack Overflow question here and close this one?

**Comment by mkonicek** [comment]
@brentvatne Agree, just added that to the docs internally, will appear on github once it lands. The JS docs don't render correctly on the website, created #2778.

**Comment by brentvatne** [comment]
Awesome :rocket:

**Comment by brentvatne** [comment]
@mkonicek @corbt - I mirrored the issue on there for you, closing this one!

**Comment by mkonicek** [comment]
Thanks @brentvatne!

**Comment by brentvatne** [comment]
https://github.com/facebook/react-native/commit/d95757037aef3fbd8bb9064e667ea4fea9e5abc1 fixed!

**Comment by lzabuaa** [comment]
Hi everybody, I wonder whether this problem has been solved in latest version? I test it in 0.24 , however  it does not work well. THX

**Comment by lzabuaa** [comment]
@brentvatne could you help me about my above question? thx a lot

**Comment by JonathanWi** [comment]
Hey guys, I'm not sure how this is solved. Below is a picture of the same code on Android (left) & iOS (right)

<img width="862" alt="capture d ecran 2016-05-31 a 17 52 51" src="https://cloud.githubusercontent.com/assets/1074479/15681208/b4319086-2758-11e6-8c77-7bec3ecac734.png">

I'm using the following code and you can obviously see the difference in execution : 

```jsx
<View style={{alignItems: 'stretch', width: 150, flexDirection: 'row', flex: 1}}>
    <View style={{backgroundColor: 'red', flex: 1, flexDirection: 'column'}}>
        <View style={{backgroundColor: 'green'}}>
          <View style={{backgroundColor: 'purple', alignSelf: 'center'}}>
            <Text>
                Column 1
            </Text>
          </View>
        </View>
      </View>

      <View style={{backgroundColor: 'blue', flex: 1, flexDirection: 'column', height: 150}}>
        <Text>Column 2</Text>
      </View>
</View>
```

How can I solve this? Is it an expected behavior? 

Thanks a lot
